x/harvest/keeper/interest.go

```diff
@@ -2,30 +2,177 @@ package keeper
+	// Get prior borrows
+	borrowsPrior := sdk.NewCoin(denom, sdk.ZeroInt())
+	borrowCoinsPrior, foundBorrowCoinsPrior := k.GetBorrowedCoins(ctx)
+	if foundBorrowCoinsPrior {
+		borrowsPrior = sdk.NewCoin(denom, borrowCoinsPrior.AmountOf(denom))
+	}
+
+	reservesPrior, foundReservesPrior := k.GetTotalReserves(ctx, denom)
+	if !foundReservesPrior {
+		newReservesPrior := sdk.NewCoin(denom, sdk.ZeroInt())
+		k.SetTotalReserves(ctx, denom, newReservesPrior)
+		reservesPrior = newReservesPrior
+	}
+
+	borrowIndexPrior, foundBorrowIndexPrior := k.GetBorrowIndex(ctx, denom)
+	if !foundBorrowIndexPrior {
+		newBorrowIndexPrior := sdk.MustNewDecFromStr("1.0")
+		k.SetBorrowIndex(ctx, denom, newBorrowIndexPrior)
+		borrowIndexPrior = newBorrowIndexPrior
+	}
+
+	// Fetch money market from the store
+	mm, found := k.GetMoneyMarket(ctx, denom)
+	if !found {
+		return sdkerrors.Wrapf(types.ErrMoneyMarketNotFound, "%s", denom)
+	}
+
+	// GetBorrowRate calculates the current interest rate based on utilization (the fraction of supply that has been borrowed)
+	borrowRateApy, err := CalculateBorrowRate(mm.InterestRateModel, sdk.NewDecFromInt(cashPrior), sdk.NewDecFromInt(borrowsPrior.Amount), sdk.NewDecFromInt(reservesPrior.Amount))
+	if err != nil {
+		return err
+	}
+
+	borrowRateSpy, err := APYToSPY(borrowRateApy)
+	if err != nil {
+		return err
+	}
+
+	interestFactor := CalculateInterestFactor(borrowRateSpy, sdk.NewInt(timeElapsed))
+	interestAccumulated := interestFactor.Mul(sdk.NewDecFromInt(borrowsPrior.Amount)).TruncateInt()
+	totalBorrowInterestAccumulated := sdk.NewCoins(sdk.NewCoin(denom, interestAccumulated))
+	totalReservesNew := reservesPrior.Add(sdk.NewCoin(denom, sdk.NewDecFromInt(interestAccumulated).Mul(mm.ReserveFactor).TruncateInt()))
+	borrowIndexNew := borrowIndexPrior.Mul(interestFactor)
+
+	k.SetBorrowIndex(ctx, denom, borrowIndexNew)
+	k.IncrementBorrowedCoins(ctx, totalBorrowInterestAccumulated)
+	k.SetTotalReserves(ctx, denom, totalReservesNew)
+	k.SetPreviousAccrualTime(ctx, denom, ctx.BlockTime())
+	return nil
+}
+
+// CalculateBorrowRate calculates the borrow rate
+func CalculateBorrowRate(model types.InterestRateModel, cash, borrows, reserves sdk.Dec) (sdk.Dec, error) {
+	utilRatio := CalculateUtilizationRatio(cash, borrows, reserves)
+
+	// Calculate normal borrow rate (under kink)
+	if utilRatio.LTE(model.Kink) {
+		return utilRatio.Mul(model.BaseMultiplier).Add(model.BaseRateAPY), nil
+	}
+
+	// Calculate jump borrow rate (over kink)
+	normalRate := model.Kink.Mul(model.BaseMultiplier).Add(model.BaseRateAPY)
+	excessUtil := utilRatio.Sub(model.Kink)
+	return excessUtil.Mul(model.JumpMultiplier).Add(normalRate), nil
+}
+
+// CalculateUtilizationRatio calculates an asset's current utilization rate
+func CalculateUtilizationRatio(cash, borrows, reserves sdk.Dec) sdk.Dec {
+	// Utilization rate is 0 when there are no borrows
+	if borrows.Equal(sdk.ZeroDec()) {
+		return sdk.ZeroDec()
+	}
+
+	totalSupply := cash.Add(borrows).Sub(reserves)
```

Comment:
The edge case: 
1. Imagine APYs are really high (like 100% APY)
2. Imagine the chain halts for a long time (like 2 years)
3. Imagine the reserve factor is really high

When the chain restarts, the accrued interest on each position would be greater than the principal, and this could create the conditions where `reserves > cash + borrows`. To handle this extremely unlikely case, I think we should check if `totalSupply` is negative and return `sdk.ZeroDec()` if it is.

Reply:
In Ethereum edge cases like these are generally handled by the SafeMath library's `add` and `sub` functions which throw an error on integer underflow and overflow of unsigned ints. I think your proposal is a good way to mimic that safeguard since sdk.Dec's `Sub` function only protects against integer overflow.

Following that thought, perhaps we should define a `SafeSub` function on sdk.Dec.

Reply:
Added in https://github.com/Kava-Labs/kava/pull/720/commits/8a251b40fae678abd83d0ec677e5b952b323f81e. The added check also resolved an issue with the 'high reserves' test case. Previously, high reserves resulted in a negative value but now the output is within the expected range.